### DISCUSSION TITLE
2FA without dedicated mobile phone/device?

### DISCUSSION THREAD

**Discussion by Retro64** [community/community#23490] [New to GitHub]
<p>Is there any consideration to have 2FA without dedicated mobile phone (e.g. via e-mail)? I mean you have 2 dedicated separate secured channels and do not necessarily need a second (physical) device. Yes, I know this way is more secured…</p>
<p>Or is there any way to use 2FA on Github without needing a dedicated second device? I am forced to use 2FA by the organization I am supporting (which is absolutly fine for me), but I am actually not willing to provide my mobile number nor installing an app bounding authentication to a specific external physical device (what might break, get lost, been forgotten at home) nor having 3 devices (computer to code, dedicated mobile device authenticated and enabled to use physical key generator via nfc).</p>

**Comment by nadiajoyce** [comment, marked as answer]
<p>Hi <a class="mention" href="/retro64">@retro64</a>,</p>
<p>Thanks for being part of the GitHub Community Forum. I’ll answer your question as best I can.</p>
<p>At this time, the only options for primary 2FA are via SMS or TOTP app. This is for a number of reasons, including security for your account. I recommend reading our article <a href="https://github.community/t/getting-started-with-github-part-3-account-security-best-practices/10196">here</a> on security for your GitHub account for more information.</p>
<p>I understand your desire to not share your mobile number and also your concerns about linking authentication to a specific device. That said, we also offer several fallback methods for 2FA, should you ever lose your device. You could set up 2FA with a TOTP app and then set a number of different fallback options, including a FIDO U2F security key. This helps provide a backup option for accessing your account should you ever lose or damage your primary device with the TOTP app installed.</p>
<p>I hope this helps! Cheers!</p>

**Reply by ideasman42** [reply]
This answer makes it seem as if you might need a mobile phone (if you've never come across a `TOTP app` before).
It turns out there are desktop TOTP applications and you don't in fact need a mobile phone for GitHub's 2FA.

I tested KeePassXC and for GitHub's 2FA, and it works well.

**Reply by SUPERcocoCats** [reply]
I do not have a mobile device, so how do I get 2fa?

**Reply by airtower-luna** [reply]
As it says in the post right above yours, and a lot of others in this thread: You can use *any* TOTP implementation.

[KeePassXC](https://keepassxc.org/) works on all common desktop systems, there are some others, libraries if you want to build your own, and even hardware tokens that can do it (if you want the security of a really separate device).

**Reply by deathybrs** [reply]
KeePassXC worked for me.  It was just the slightest bit fiddly, but nothing too difficult.

**Reply by mwpowellhtx** [reply]
👍 _KeePassXC_, for now, unless until anything better comes along.

**Reply by perkinslr** [reply]
If you are running Linux, `oathtool` is the simplest solution.  No major disruption to your workflow, and it can be easily integrated into the standard unix password store, or whatever system you already use (or you can memorize the GH secret, it's short and simple).

**Comment by Retro64** [comment]
<p>Hi <a class="mention" href="/nadiajoyce">@nadiajoyce</a>,</p>
<p>thank you for your answer (even if it was not the answer I hoped for…). I still hope there will be a multi channel, one device solution for 2FA in the near future…</p>

**Comment by nhatkhai** [comment]
<p>I’m hate this nonsese crap of MFA. Company still get hack from inside out anyway! I also not rich enough to own a smart phone, or paid for SMS crap.</p>

**Comment by nhatkhai** [comment]
<p>Please give us the users freedome to by pass this crap without smart phone and SMS.</p>

**Comment by redhatdragon** [comment]
<p>To be honest MFA using phones or any second device in general is a very bad idea.  Besides the fact that some of us are too poor to own say a smartphone (myself included) it just increases the attack vector and doesn’t solve really anything, and encourages data loss especially among people who wouldn’t even bother to use a strong password in the first place.  There’s no excuse for this.  It’s a very bad idea.  And immoral to force it on people.  As of current I can’t twitch stream, can hardly use steam, can’t really use any service which depends on it.  The only reason I have an email is because thank god there’s some email services that don’t require a **bleep** phone number.  I could see it being optional.  But making it mandatory is just very against personal rights.</p>

**Comment by sparker-dg** [comment]
<p>Posted in wrong “reply” location. Will go recreate in appropriate location.</p>

**Comment by airtower-luna** [comment]
<p>I haven’t tested it with Github 2FA, but <a href="https://keepassxc.org/">KeePassXC</a> is a desktop password manager that can also do TOTP. Carefully consider whether keeping the TOTP tool on (presumably) the same computer that you use to log in fits your security needs.</p>

**Comment by JustKuzya** [comment]
<p>I don’t own any cell-network device. Does it mean that I’ll never be able to use my github account I use now?</p>

**Comment by airtower-luna** [comment]
<ol>
<li>You do not have to use 2FA.</li>
<li>If you want to use 2FA, you do not have to use SMS. TOTP is safer (because SMS might be intercepted) and doesn’t require a mobile connection.</li>
</ol>

**Reply by michaelerule** [reply]
No longer true as of 2023---Github has been mandating that anyone who "contributes code" enable 2FA, or else lose access to their account. I'm currently tying to figure how how this can be done robustly without a phone.

**Reply by mwpowellhtx** [reply]
_Keepassxc_ ...

**Reply by deathybrs** [reply]
While I am not entirely pleased at the requirement, I do want to back up @mwpowellhtx's post - KeePassXC worked for me.  It is F/OSS, and runs on the desktop.

**Reply by michaelerule** [reply]
Thanks; I'm trying to navigate this process, and am finding the documentation lacking. Github provides no instructions for working with keepassxc, and keepassxc's own "getting started" workflows don't immediately/obviously provide clues for how to get past Github's QR code 2FA start page. I'm still looking for a clear tutorial on how keepassxc+github *actually* works, and will post any links here if I can find them.

**Reply by michaelerule** [reply]
Okay, the sticking point was just the `keepassxc` GUI. Graphical instructions are [here](https://keepassxc.org/docs/KeePassXC_UserGuide), `ctl+f` for words like *paste* and *secret*

Steps I followed: 
 
 - Install `keepassxc` from a PPA
```
sudo add-apt-repository ppa:phoerious/keepassxc
sudo apt update
sudo apt install keepassxc 
```
 - Install the [Firefox plugin](https://addons.mozilla.org/en-US/firefox/addon/keepassxc-browser/)
 - Run `keepassxc`
   - Create a new database (I followed default settings)
   - Unlock database (padlock icon at top of keepassxc GUI)
   - Go to settings (gear icon) -> browser integration
   - Enable browser integration (top check box), and then enable integration for firefox (check box within tab)
 - From Firefox, activate `keepassxc` extension (extension manager puzzle icon in top right of window) and click "connect"
   - Navigate to Githubs 2FA setup page with the QR code
   - Click to reveal the "secret" as a base-64 text string, rather than QR code; Copy this code. 
 - Return to  `keepassxc` GUI
   - Add a new entry (`+` icon at the top of the window)
   - Add some details here (although I'm not sure they matter---beyond giving this entry a name that helps you remember its for github)
   - Exit/complete the entry configuration and switch back to the list-view of password database entries
   - Right click on the new entry for github, and select "add TOPT"---this seems to be the *only* place where this functionality is exposed, it's not accessible within the tabs/GUI when you click to edit an entry
   - Paste in the "secret" from github's 2FA start page, and click to finish. 
   - In the list view of your password database in keepassxc, you should now see a (clock) symbol to the left, in the entry row for github
   - Click on this clock. Nothing will happen at first, but your TOPT code has been copied to the clipboard. Notice that this starts a 10 second count-down, at the bottom of the keepassxc GUI, which will clear the clipboard. 
   - Paste your TOPT into the text field provided on the Github 2FA start page and proceed. 

Be sure to save/print your recovery codes provided by the Github website after finishing these steps , (and quickly); My understanding is one will be permanently locked out of one's account if the second factor device fails, and these codes are lost. 

Next step: Setting this up on more than one physical device to be robust to hardware failures. 

Thanks! ~m

**Reply by michaelerule** [reply]
Ok, stupid question: How do I set this up so that I can log in from either my desktop at home or at the office? I've set up `keepasscx` on one desktop. But, I now need to retrieve the TOPT somehow whenever I log in. I can do this using remote access to/from my pair of desktops, but it seems more robust to configure *multiple* authenticator apps, any one of which would suffice. Unfortunately, Github's 2FA settings allow you to *reconfigure* an authenticator app (presumably voiding the old one), but not to add a *second* TOPT authenticator app+device as a backup. Is this possible and if so where might one find detailed instructions? 

#### Resolution: 

The "two-factor secret" encoded in the QR code when you initially set up Github 2FA can be used to seed TOPT using keepassxc on multiple devices. If you've saved the seed (the text you get when you click to view secret, under the QR code), this is all you need to set up multiple devices. It seems recommended to store this secret as physical copie(s) in a safe place. Access to this seed, alongside one's account details, will allow anyone to create a new 2FA authentication device.

**Reply by airtower-luna** [reply]
You can also have KeePassXC display the QR code, or look up the secret in the advanced attributes of the entry, so there's no need to save it separately (which then would raise the question of how to protect that). And if you use KeePassXC on multiple devices you can just copy the database between them. :slightly_smiling_face:

**Reply by mwpowellhtx** [reply]
I use the combination _KeePassXC_ and _ShareX_. _ShareX_ pretty powerful, not only can you compose (_encode_) your own QR codes, but also _decode_ them.

**Reply by michaelerule** [reply]
@airtower-luna

Just a note, I recently attempted this via database copying and encountered issues. The problem ended up just being clock/time skew, and re-syncing the system times fixed it.

**Reply by airtower-luna** [reply]
That makes sense, being time-based (the first T) TOTP needs mostly accurate time to work right. :slightly_smiling_face:

**Comment by myudkowsky** [comment]
<p>You can also, for a nominal monthly fee, get a phone number at a provider, e.g., voip.ms. That phone number, which does not ever have to be used to send or receive phone calls, can forward you SMS messages via email, and you can send messages as well.</p>

**Reply by mwpowellhtx** [reply]
well again as others have stated very clearly, nominally ANY monthly fee is UNACCEPTABLE.

**Comment by ideasman42** [comment]
<p>Any updates on this? I don’t own a phone and apparently 2FA will be enforced eventually. <a href="https://venturebeat.com/2022/05/04/github-to-enforce-2fa-for-all-code-contributors-by-the-end-of-2023/" class="inline-onebox" rel="noopener nofollow ugc">GitHub to enforce 2FA for all code contributors by the end of 2023 | VentureBeat</a></p>
<p>Are there any known working options for 2FA such as USB/smart-card that don’t require a phone?</p>

**Comment by airtower-luna** [comment]
<p>As I <a href="https://github.community/t/2fa-without-dedicated-mobile-phone-device/835/9">mentioned above</a> you can use something that implements TOTP on your desktop, like KeePassXC. No phone needed. :slightly_smiling_face:</p>

**Reply by mwpowellhtx** [reply]
will look into that, thank you...

**Comment by ideasman42** [comment]
<p>Thanks, KeePassXC works great!</p>
<p>I was confused by the wording on github’s 2FA page which made it seems like I might need a phone (use of QR code for e.g.).</p>
<p>But in fact it’s not needed at all.</p>

**Reply by mwpowellhtx** [reply]
works well enough and with complimentary QR code tooling...

**Comment by thelllmike** [comment]
Yes, GitHub supports using a security key (like a YubiKey) for 2FA without requiring a dedicated mobile phone or app. This method enhances security and doesn't rely on a second physical device.

**Reply by ideasman42** [reply]
Are you sure about this?

While a security-key is a 2nd factor, if I try to enable 2FA (the thing which will be mandatory in the future apparently). The only option is to use SMS / TOTP-app (QR-code etc).

There is no option to only use the security key (USB-security-token).

**Reply by mwpowellhtx** [reply]
90 percent certain, yes, I seem to recall that being one of the options. How it is actually facilitated, could not tell you.

**Reply by ideasman42** [reply]
It would be good if this were the case, I use a Nitrokey with Github (to push), so it would be great if this could _be_ the 2nd factor, but I tried again and enabling 2FA requires either a:

- QR-CODE.
- authenticator app (which where I can enter the code given).
- SMS

There is no option to use a USB security token from what I can tell.

**Reply by mwpowellhtx** [reply]
This is what I see for my options. Not sure what your view enumerates. But if that is to do with _USB security keys_, then perhaps. _TL;DR_, I have not explored that feature, because for now `KeePassXC` seems adequate.

![image](https://github.com/community/community/assets/1824187/936196fc-949b-4efe-92f3-0038069f7448)

**Reply by ideasman42** [reply]
This option looks only to be available **after** you configure 2FA,see: 

https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#configuring-two-factor-authentication-using-a-security-key

> After you configure 2FA, using a time-based one-time password (TOTP) mobile app, or via text message, you can add a security key, like a fingerprint reader or Windows Hello. The technology that enables authentication with a security key is called [WebAuthn](https://webauthn.guide/). WebAuthn is the successor to U2F and works in all modern browsers.

**Reply by mwpowellhtx** [reply]
Not required to have a mobile phone to clear the 2FA hurdle.

**Comment by nhatkhai** [comment]
For me sending the verify code to email already is multi-factor-authentication. Since I use private mode browser all most all the time, entering the code also all most all the time. There are **no op-out option at all**.... It like some one got hacked, and become everyone else duty to do extra work! Sound very non-green.

**Reply by mwpowellhtx** [reply]
I do not disagree; however, those options do not appear to be readily available. Just working with what I do have.

**Comment by deathybrs** [comment]
So, what I'm hearing is that it's time to move somewhere else if we don't want to worry about permanently losing access to our code.

Any suggestions?

**Reply by ideasman42** [reply]
Checkout: https://codeberg.org (a hosted [Forgejo](https://forgejo.org) instance - a fork of Gitea).

Unlike github, you can use 2FA via a USB-security token without a mobile-phone or a TOTP-app (which I found a hassle to setup & use).

**Reply by mwpowellhtx** [reply]
That's one choice, of course, I don't think anyone is 'forcing' you to stay here, although the polices in effect are weeding folks in or out, by design I think. Git being what it is, repo can live 'anywhere'. I've used bitbucket before, which is fine for small sometimes proivate projects as well.

**Comment by sparker-dg** [comment]
One thing to note here in this thread about the reliance on a physical device as your TOTP generator. If you save the 2FA serial number you receive when setting up your 2FA device, you can plug that into a number of different devices and applications to generate the same keys. I never set up a 2FA without recording the serial number in my password app. That way, if I lose a phone, or Google Authenticator is discontinued, or Authy is discontinued, I simply plug that serial number into a new TOTP generator, and have exactly the same codes available as my original device. 

1Password has an awesome method of storing the serial number that actually generates the codes directly, so if I have my phone handy, I can use it for 2FA. If I don't, I can get my codes from 1Password. And on a couple other systems, I have Authy available. All of them generate the same codes. 

Might be worth considering using multiple TOTP generators as a solution.

**Comment by wothke** [comment]
Stop spamming my email with your 2FA reminders! I will not use that 2FA garbage and I already moved all my stuff to a more user friendly git hoster. You can stick that garbage where the sun doesn't shine...  

and your idea of "Contact Us" is a bloody chat bot? FU!

**Comment by nhatkhai** [comment]
I trust my own password, and I keep in my brain. Are you GitHub ?

**Comment by objectkit** [comment]
Unless I can find a white hat solution to this, it's time for me to leave the tiny little neurotypical world of microsoft. Wishing you all peace, happiness and prosperity without intermediary. 

All the best.

**Comment by surfskidude** [comment]
My religion does not allow me to use a phone. What do I do?

**Reply by mwpowellhtx** [reply]
Ad nauseam `KeePassXC` seems to work in most cases.

**Reply by ideasman42** [reply]
> My religion does not allow me to use a phone. What do I do?

What I've found works well on Linux:
- KeePassXC for the initial 2FA setup.
- A USB security token (such as a NitroKey3) which works as a passkey.

This makes logging-in more or less usable. Using KeePassXC regularly is an impractical hassle.

**Reply by michaelerule** [reply]
Indeed, KeePassXC works. Well, on some of my linux installs its UI lacks text (a bug affecting all QT apps I've yet to resolve). But it does the thing.

**Reply by airtower-luna** [reply]
> Using KeePassXC regularly is an impractical hassle.

What's impractical about it? With the matching browser extension filling the login code takes only one click.

**Reply by ideasman42** [reply]
> > Using KeePassXC regularly is an impractical hassle.
> 
> What's impractical about it? With the matching browser extension filling the login code takes only one click.

This is the first I've heard of the browser extension, maybe that resolves the hassles. In general though I find KeePassXC annoying to have to interact with, it has it's own password and local configuration which you can't easily move your setup between systems.
You need KeePassXC, & the browser extension & the local configuration... maybe OK but I'd rather avoid the overhead of having to have this setup on my systems.

**Reply by mwpowellhtx** [reply]
So does your smart device, either a pass code, pattern, bio metric, etc, and if you don't own, or will not, for whatever reason. The whole MFA challenge is 'something you own' with reasonable security measures. If you have a better idea pitch it. 💡 🍻

**Reply by airtower-luna** [reply]
> it has it's own password and local configuration which you can't easily move your setup between systems.

You can, it's as easy as copying (or synchronizing) the DB file. :wink: It has some extra sync support if you make changes on multiple systems without synchronizing first, but I haven't needed that so far.

If you use the browser extension you'll have to pair the DB file with each profile you use, but as long as you synchronize that back that's a one time thing. As long as you don't switch/factory-reset your browser, of course.

**Reply by surfskidude** [reply]
> Ad nauseam `KeePassXC` seems to work in most cases.
I installed it on my Win 11 machine, but it's not starting. It may have crashed, but I did not see any errors.
